File: crud-jpa/src/main/java/me/jiangcai/crud/row/field/FieldBuilder.java
package me.jiangcai.crud.row.field;

import me.jiangcai.crud.row.FieldDefinition;
import me.jiangcai.crud.row.field.fake.AbstractFake;
import me.jiangcai.crud.row.field.fake.FakeRoot;
import org.springframework.beans.BeanUtils;
import org.springframework.http.MediaType;

import javax.persistence.criteria.*;
import java.lang.reflect.InvocationTargetException;
import java.util.List;
import java.util.function.BiFunction;
import java.util.function.Function;

/**
 * 字段构建器
 *
 * @author CJ
 */
public class FieldBuilder<T> {

    private final String name;
    private BiFunction<Root<T>, CriteriaBuilder, Expression<?>> biSelect;
    private OwnExpression<T> ownSelect;
    private Function<Root<T>, Expression<?>> select;
    private BiFunction<Object, MediaType, Object> format;
    private Function<Root<T>, Expression<?>> order;
    private BiFunction<Root<T>, CriteriaBuilder, Expression<?>> biOrder;
    private Function<T, ?> entityFunction;
    private boolean noOrder = false;
    private OwnExpression<T> ownOrder;

    private FieldBuilder(String name) {
        this.name = name;
    }

    /**
     * 开始一个构造器
     *
     * @param name {@link FieldDefinition#name()}
     * @param <Y>  动态类型
     * @return 新的构造器
     */
    public static <Y> FieldBuilder<Y> asName(Class<Y> type, String name) {
        return new FieldBuilder<>(name);
    }

    /**
     * 开始一个构造器
     *
     * @param name {@link FieldDefinition#name()}
     * @return 新的构造器
     */
    public static <X> FieldBuilder<X> asName(String name) {
        return new FieldBuilder<X>(name);
    }

    /**
     * 获取数据若是需要CriteriaBuilder的配合
     *
     * @param function 支持传入 CriteriaBuilder的定制函数
     * @return this
     */
    public FieldBuilder<T> addBiSelect(BiFunction<Root<T>, CriteriaBuilder, Expression<?>> function) {
        this.biSelect = function;
        return this;
    }

    /**
     * 获取数据时若需要当时的查询配合
     *
     * @param ownSelect 完全定制的函数
     * @return this
     */
    public FieldBuilder<T> addOwnSelect(OwnExpression<T> ownSelect) {
        this.ownSelect = ownSelect;
        return this;
    }

    /**
     * 获取数据仅需Root
     *
     * @param function root为参数的函数
     * @return this
     */
    public FieldBuilder<T> addSelect(Function<Root<T>, Expression<?>> function) {
        this.select = function;
        return this;
    }

    public FieldBuilder<T> addOwnOrder(OwnExpression<T> ownOrder) {
        this.ownOrder = ownOrder;
        return this;
    }

    public FieldBuilder<T> addBiOrder(BiFunction<Root<T>, CriteriaBuilder, Expression<?>> function) {
        this.biOrder = function;
        return this;
    }

    public FieldBuilder<T> addOrder(Function<Root<T>, Expression<?>> function) {
        this.order = function;
        return this;
    }

    public FieldBuilder<T> withoutOrder() {
        this.noOrder = true;
        return this;
    }

    /**
     * 从实体中获取的数据并不符合数据规格，则可调用该方法
     *
     * @param format 将原数据和期望Media转变为新数据
     * @return this
     */
    public FieldBuilder<T> addFormat(BiFunction<Object, MediaType, Object> format) {
        this.format = format;
        return this;
    }

    /**
     * 如果采用了{@link #addBiSelect(BiFunction)}或者{@link #addOwnSelect(OwnExpression)}会导致生成的FieldDefinition的
     * {@link FieldDefinition#readValue(Object)}无法正常工作，这个时候需要调用该方法实施定制
     *
     * @param entityFunction 获取值的办法
     * @return this
     */
    public FieldBuilder<T> addEntityFunction(Function<T, ?> entityFunction) {
        this.entityFunction = entityFunction;
        return this;
    }

    public FieldDefinition<T> build() {
        return new FieldDefinition<T>() {
            @Override
            public Object readValue(T entity) {
                if (entityFunction != null)
                    return entityFunction.apply(entity);
                if (select != null) {
                    AbstractFake fake = (AbstractFake) select.apply(new FakeRoot<>());
                    return fake.toValue(entity);
                }
                try {
                    return BeanUtils.getPropertyDescriptor(entity.getClass(), name()).getReadMethod().invoke(entity);
                } catch (IllegalAccessException | InvocationTargetException e) {
                    throw new IllegalStateException(e);
                }
            }

            @Override
            public Selection<?> select(CriteriaBuilder criteriaBuilder, CriteriaQuery<?> query, Root<T> root) {
                if (ownSelect != null)
                    return ownSelect.toExpression(root, criteriaBuilder, query);
                if (biSelect != null)
                    return biSelect.apply(root, criteriaBuilder);
                if (select != null)
                    return select.apply(root);
                return root.get(name);
            }

            @Override
            public String name() {
                return name;
            }

            @Override
            public Object export(Object origin, MediaType mediaType, Function<List, ?> exportMe) {
                if (format != null)
                    return format.apply(origin, mediaType);
                return origin;
            }

            @Override
            public Expression<?> order(CriteriaQuery query, CriteriaBuilder criteriaBuilder, Root<T> root) {
                if (noOrder)
                    return null;
                if (ownOrder != null)
                    return ownOrder.toExpression(root, criteriaBuilder, query);
                if (biOrder != null)
                    return biOrder.apply(root, criteriaBuilder);
                if (order != null)
                    return order.apply(root);
                if (ownSelect != null)
                    return ownSelect.toExpression(root, criteriaBuilder, query);
                if (biSelect != null)
                    return biSelect.apply(root, criteriaBuilder);
                if (select != null)
                    return select.apply(root);
                return root.get(name);
            }
        };
    }

    @FunctionalInterface
    public interface OwnExpression<T> {
        Expression<?> toExpression(Root<T> root, CriteriaBuilder cb, CriteriaQuery<?> query);
    }


}

File: crud-jpa/src/main/kotlin/me/jiangcai/crud/row/FieldBuilder.kt
package me.jiangcai.crud.row

import me.jiangcai.crud.row.field.fake.AbstractFake
import me.jiangcai.crud.row.field.fake.FakeCriteriaBuilder
import me.jiangcai.crud.row.field.fake.FakeCriteriaQuery
import me.jiangcai.crud.row.field.fake.FakeRoot
import org.springframework.beans.BeanUtils
import org.springframework.core.convert.ConversionService
import org.springframework.http.MediaType
import java.lang.reflect.InvocationTargetException
import java.util.function.Function
import javax.persistence.criteria.*
import javax.persistence.metamodel.SingularAttribute

typealias ToSelect<X, T> = (From<*, T>, CriteriaBuilder, CriteriaQuery<*>) -> Expression<out X>
typealias ToFormat = (Any?, MediaType?, exportMe: Function<MutableList<Any?>, *>?) -> Any?

/**
 * 新版本构建器。
 * @param T 当前build类型
 * @param S 原始Entity类型
 * @author CJ
 */
class FieldBuilder<T, S>(
    private val type: Class<S>,
    private val conversionService: ConversionService,
    val toReadFrom: (From<*, S>) -> From<*, T>
) {
    fun <Y> forBuilder(from: (From<*, T>) -> From<*, Y>): FieldBuilder<Y, S> {
        return FieldBuilder(
            type, conversionService
        ) { from(toReadFrom(it)) }
    }

    /**
     * @return 一个[ConversionService]友好的格式工具
     */
    @Suppress("unused")
    fun formatVia(conversionService: ConversionService = this.conversionService): ToFormat {
        return { value, _, _ ->
            if (value == null) null
            else
                conversionService.convert(value, String::class.java)
        }
    }

    /**
     * 直接用实体的字段名称构建
     * @param field 字段名称
     * @param name 如同[FieldDefinition.name],默认就是字段名称
     * @param order 是否支持排序, 更多细节遵守[RowDramatizer.order]
     * @param format 自定义格式化
     */
    inline fun <reified X> forField(
        field: String,
        name: String = field,
        order: Boolean = true,
        noinline format: ToFormat = { input, _, _ -> input }
    ): TypeFieldDefinition<X, S> {
        return MyField(
            name = name,
            resultType = X::class.java,
            order = order,
            selector = { root, _, _ -> toReadFrom(root).get(field) },
            format = format
        )
    }

    /**
     * 支持更富有想象力的字段
     *
     * @param name 如同[FieldDefinition.name]
     * @param selector 如何从持久层中获取需要的数据
     * @param order 是否支持排序, 更多细节遵守[RowDramatizer.order]
     * @param format 自定义格式化
     */
    inline fun <reified X> forSelect(
        name: String,
        noinline selector: ToSelect<X, T>,
        order: Boolean = true,
        noinline format: ToFormat = { input, _, _ -> input }
    ): TypeFieldDefinition<X, S> {
        return MyField(
            name = name,
            resultType = X::class.java,
            order = order,
            selector = { from, cb, cq ->
                selector(toReadFrom(from), cb, cq)
            },
            format = format
        )
    }

    /**
     * 使用jpa metamodel 属性构建
     *
     * @param attribute metamodel 属性
     * @param name 如同[FieldDefinition.name],默认就是字段名称
     * @param order 是否支持排序, 更多细节遵守[RowDramatizer.order]
     * @param format 自定义格式化
     */
    fun <X> forAttribute(
        attribute: SingularAttribute<in T, X>,
        name: String = attribute.name,
        order: Boolean = true,
        format: ToFormat = { input, _, _ -> input }
    ): TypeFieldDefinition<X, S> {
        return MyField(
            name = name,
            resultType = attribute.javaType,
            order = order,
            selector = { root, _, _ -> toReadFrom(root).get(attribute) },
            format = format
        )
    }

    inner class MyField<X>(
        private val name: String,
        private val resultType: Class<X>,
        private val selector: ToSelect<X, S>,
        private val order: Boolean,
        private val format: ToFormat
    ) : TypeFieldDefinition<X, S> {
        override fun getResultType(): Class<X> = resultType

        override fun select(cb: CriteriaBuilder, query: CriteriaQuery<*>, root: Root<S>): Expression<out X> {
            return selector(root, cb, query)
        }

        override fun export(origin: Any?, mediaType: MediaType?, exportMe: Function<MutableList<Any?>, *>?): Any? {
            return format(origin, mediaType, exportMe)
        }

        override fun order(
            query: CriteriaQuery<*>,
            criteriaBuilder: CriteriaBuilder,
            root: Root<S>
        ): Expression<*>? {
            if (!order)
                return null
            return select(criteriaBuilder, query, root)
        }

        override fun name(): String = name

        override fun readValue(entity: S): Any {
//        if (entityFunction != null)
//            return entityFunction.apply(entity)
            return try {
                val fake = select(FakeCriteriaBuilder(), FakeCriteriaQuery(), FakeRoot<S>()) as AbstractFake
                fake.toValue(entity)
            } catch (e: Throwable) {
                try {
                    BeanUtils.getPropertyDescriptor(type, name()).readMethod.invoke(entity)
                } catch (e: IllegalAccessException) {
                    throw IllegalStateException(e)
                } catch (e: InvocationTargetException) {
                    throw IllegalStateException(e)
                }
            }


//        return super.readValue(entity)
        }
    }
}

